Update `LogMergePolicy` to skip to a target number of documents

kryesh opened the pull request:
From discord conversation:
https://discord.com/channels/908281611840282624/915785344396439552/1341705839668625468

This updated logic makes `LogMergePolicy` aim for a specific target number of documents, and opportunistically skip merge operations to reach that target document count.
Pros:
* Reduced IO/CPU usage from skipping intermediate merge operations
* No longer susceptible to creating huge merge operations by merging many large segments into a single segment many times larger than the target size (previously `max_docs_before_merge`)
  *  The theoretical maximum size of a segment with this updated logic is `(target_segment_size * 2) - 2`
Cons:
*  If an index has a little over `target_segment_size` total docs then it may get merged to a single segment and thus not parallelize well

fulmicoton commented on src/indexer/log_merge_policy.rs:
```diff
@@ -76,62 +76,97 @@ impl LogMergePolicy {
+                    batch_docs = 0;
+                    candidates.push(MergeCandidate(
+                        // drain to reuse the buffer
+                        batch.drain(..).map(|seg| seg.id()).collect(),
```
Nothing prevents this merge to have 1000 segments.

fulmicoton commented on src/indexer/log_merge_policy.rs:
```diff
@@ -76,62 +76,97 @@ impl LogMergePolicy {
-            .collect()
+            .into_iter()
+            .for_each(|(_, group)| {
+                let mut hit_delete_threshold = false;
```
for loop would be nicer than for_each here.

fulmicoton commented on src/indexer/log_merge_policy.rs:
```diff
@@ -76,62 +76,97 @@ impl LogMergePolicy {
+                    // If there aren't enough documents to create another segment of the target size
+                    // then break
+                    if unmerged_docs <= self.target_segment_size {
+                        break;
```
aaaaaaaaaaaarg the original merge policy has the same flaw!

kryesh commented on src/indexer/log_merge_policy.rs:
```diff
@@ -76,62 +76,97 @@ impl LogMergePolicy {
+            .into_iter()
+            .for_each(|(_, group)| {
+                let mut hit_delete_threshold = false;
+                group.for_each(|(_, seg)| {
```
as above, will update

stuhood commented:
> @stuhood Can you review? Datadog/Quickwit does not use this merge policy.

Mm, yea: we don't either.

----

It looks like the intent of your change is to avoid repeatedly re-merging very small segments. Unlike an LSM-tree, Tantivy does not have a "memtable", and so there is no minimum segment size, and very small segments might be created. `LogMergePolicy::clip_min_size` can cause these very small segments to be merged repeatedly, because it effectively lies about the actual size of a small segment. If my math is right, due to `clip_min_size`, a segment containing 1 document might be merged 5 times before it reaches the `DEFAULT_MIN_LAYER_SIZE`.

As @fulmicoton pointed out, the effect of your change might be to merge an arbitrary number of segments (1000 etc). But probably more importantly, it might merge segments of drastically different sizes: imagine that deletes are arriving which shrink segments which were _larger_ than your `target_segment_size` to _smaller_ than your `target_segment_size` (i.e. from 1.01GB to 0.99GB for a target of 1GB). Now that much larger segment is eligible to be merged with potentially super tiny segments: if some brand new segments arrive which are only 10MB, they might end up being compacted with your (now) 0.99GB segment. And this might happen repeatedly.

This is why the log scale exists: it causes extra [write amplification](https://www.scylladb.com/2018/01/31/compaction-series-leveled-compaction/) to merge tiny segments with larger segments unnecessarily. A `target_segment_size` which is too high, _or_ the `clip_min_size` behavior can cause lots of redundant merging.

I like the general thrust of this PR, but I think that what I would propose would be to:
* remove the `min_layer_size` and the `clip_min_size` behavior
  * as discussed above: they lead to write amplification
* significantly lower your new `target_segment_size` (to e.g. 100 or 500 documents), and rename it to `target_min_segment_size` to replace the previous intent of the `min_layer_size` settings
  * this value would act like "the memtable size" in an LSM tree: it would be the initial value in the lowest level of the log-scale behavior
  * when this value is lower, the consequences of the case I described above (deletes causing a segment to wobble back and forth around the target) are less severe
* remove the `max_docs_before_merge` setting, and instead allow the log scale to be applied to all segments which are larger than `target_min_segment_size`
  * AFAICT, the `max_docs_before_merge` setting used to represent a ceiling on which segments would be merged. but I don't think that this is a good idea, for a few reasons:
    1. term dictionaries must be consulted and decompressed in every segment, which has overhead. doing this fewer times is preferable
    2. metadata has a cost: having thousands of tiny segments increases the cost of storing and loading metadata

stuhood commented:
> remove the `min_layer_size` and the `clip_min_size` behavior ... significantly lower your new `target_segment_size` (to e.g. 100 or 500 documents), and rename it to `target_min_segment_size` to replace the previous intent of the `min_layer_size` settings 

...or, alternatively: don't add `target_segment(_min)_size`, and instead lower the `min_layer_size` to 100 or 500 docs to make the write amplification there less severe. Would still recommend removing `max_docs_before_merge`.

kryesh commented:
The main issue I was encountered wasn't the very small segments, but repeated merges of intermediate/large segments which take a lot longer and block other operations - but yes the intent was to prevent repeated merges for the same documents.

Regarding deletes and having a segment flapping between just above and just below the target size, would changing the default delete ratio to something like 30% then updating the filter logic to look at a combination of max_doc and the delete ratio for the segment make more sense? Possibly merge segments with either of this criteria instead of num_docs below the target?:
- max_doc below the target
- max_doc above the target + hit the delete ratio threshold

The min size/clipping behaviour is a holdover from the previous policy, I think it'd be worth testing with it removed and see how it reduces write amplification. I can look into this.

Removing the segment limit/target is nice in theory like you mention with the term dictionaries etc - but there are practical issues that prevent that from working such as the 32 bit doc address value imposing a hard limit, #2577 is one example of this. There needs to be some form of maximum size enforced since not having a limit leads to other problems with huge merge operations which was another goal of this PR. If anything making the default target size bigger to approach the practical limit might make more sense, 200m or so?

stuhood commented:
> Removing the segment limit/target is nice in theory like you mention with the term dictionaries etc - but there are practical issues that prevent that from working such as the 32 bit doc address value imposing a hard limit, #2577 is one example of this. There needs to be some form of maximum size enforced since not having a limit leads to other problems with huge merge operations which was another goal of this PR.

Yea, true. But `const DEFAULT_MAX_DOCS_BEFORE_MERGE: usize = 10_000_000;` is overkill for accomplishing that. Could massively increase it.

kryesh commented:
@stuhood  @fulmicoton Pushed some more updates to address the feedback so far.


>As @fulmicoton pointed out, the effect of your change might be to merge an arbitrary number of segments (1000 etc). But probably more importantly, it might merge segments of drastically different sizes: imagine that deletes are arriving which shrink segments which were larger than your target_segment_size to smaller than your target_segment_size (i.e. from 1.01GB to 0.99GB for a target of 1GB). Now that much larger segment is eligible to be merged with potentially super tiny segments: if some brand new segments arrive which are only 10MB, they might end up being compacted with your (now) 0.99GB segment. And this might happen repeatedly.

To address this I updated the filter logic at the start to be a bit smarter than naively using `num_docs`, with this change borderline segments that have some documents deleted won't immediately trigger a merge until a certain ratio is reached. I also updated the default delete ratio - but I'm not sure what the "backward compatibility" comment on the current version is referring to.

Another leftover behaviour from the current policy is this chunk which forces a merge if any segment has hit the delete threshold.
```rust
let mut hit_delete_threshold = false;
for (_, segment) in group {
    batch.push(segment.id());
    hit_delete_threshold |= self.segment_above_deletes_threshold(segment);
}

if batch.len() >= self.min_num_segments || hit_delete_threshold {
```
should this still be a thing? or just use `num_docs` like all the other segments and remove this check? Not sure if there's a historic reason for this or if deletes are really bad for performance or something so figured I'd ask.

fulmicoton commented:
Not the scope of this PR, but here is my thoughts on what a simple rewrite of the log merge policy could look like.
https://github.com/quickwit-oss/tantivy/issues/2794

stuhood reviewed (commented):
Changing the `DEFAULT_MIN_LAYER_SIZE` gets this much closer I think.

My primary concern though is that it adds a _lot_ of code that effectively diverges from the name of the policy: about 100 lines dedicated to something that sidesteps the "Log" part of LogMerge.

> The main issue I was encountered wasn't the very small segments, but repeated merges of intermediate/large segments which take a lot longer and block other operations - but yes the intent was to prevent repeated merges for the same documents.

@kryesh : If that's the case, and it's actually medium/large segments that you're worried about, then perhaps what you want to do is significantly increase the `min_num_segments` value (e.g. to 32), so that merges happen much less frequently (and more segments are alive at a time)?

With the new `DEFAULT_TARGET_SEGMENT_SIZE = 200_000_000` value that you've set, my concern is that your "target" size is now just a strange sort of "alternate level" which isn't actually computed from the log scale. It is _not_ doing what we discussed in https://github.com/quickwit-oss/tantivy/pull/2627#issuecomment-3708798652 (acting as a cap on max segment size): that would be something else.

kryesh commented:
Fair, but these options are exposed to users anyway so in my case I'd just use my current target setting of ~33m which I've found works well. The value of 200m means there is an effective hard cap of 400m (399_999_998) docs per segment.
This PR was always adding an alternate level that bypasses the log scale since the current log algorithm isn't capable of enforcing a maximum size.
